### PR TITLE
Fn/refactoring

### PR DESCRIPTION
**Pull request by felipenoris**
This adds some love to the package, porting to Julia v0.6. After this, I can work on v1.0.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,9 @@
-*.cov
+*.jl.cov
+*.jl.*.cov
+*.jl.mem
+/deps/build.log
+/deps/deps.jl
+/deps/usr/
+/docs/build/
+/docs/site/
+/docs/out.xml
```
Comment (ScottPJones) [inline .gitignore]: should have trailing \n, some tools don't like not having it

File: .travis.yml [hunk 2]
```diff
@@ -1,17 +1,11 @@
...
-  - if [[ -a .git/shallow ]]; then git fetch --unshallow; fi
-script:
-  - julia -e 'Pkg.init()'
-  - julia -e 'Pkg.clone(pwd()); Pkg.build("LibBSON")'
-  - julia -e 'Pkg.test("LibBSON", coverage=true)'
+notifications:
+    email: false
 after_success:
-- julia -e 'cd(Pkg.dir("LibBSON")); Pkg.add("Coverage"); using Coverage; Coveralls.submit(Coveralls.process_folder())'
+  - julia -e 'Pkg.add("Coverage"); cd(Pkg.dir("LibBSON")); using Coverage; Coveralls.submit(Coveralls.process_folder())'
```
Comment (ScottPJones) [inline .travis.yml]: These can be combined:
```yaml
  - julia -e 'VERSION < v"0.7.0-DEV" || (using Pkg); cd(Pkg.dir("LibBSON")); Pkg.add("Coverage"); using Coverage; Coveralls.submit(Coveralls.process_folder()); Codecov.submit(Codecov.process_folder())'
```

File: REQUIRE [hunk 3]
```diff
@@ -1,4 +1,2 @@
-julia 0.5
-BinDeps 0.4
-Compat 0.9
-@osx Homebrew
+julia 0.6
```
Comment (ScottPJones) [inline REQUIRE]: I'd cap this at 1.0 (or 0.7, if it doesn't work on 0.7)

File: src/BSONIter.jl [hunk 4]
```diff
@@ -97,35 +98,29 @@ end
 
 # Iterator
 
-start(bsonObject::BSONObject) = begin
+Base.start(bsonObject::BSONObject) = begin
```
Comment (ScottPJones) [inline src/BSONIter.jl]: I'd get rid of the begin return end, i.e.:
`Base.start(obj::BSONObject) = BSONIter(obj)`

File: src/BSONIter.jl [hunk 5]
```diff
@@ -97,35 +98,29 @@ end
...
     return BSONIter(bsonObject)
 end
-export start
 
-next(bsonObject::BSONObject, bsonIter::BSONIter) = begin
+Base.next(bsonObject::BSONObject, bsonIter::BSONIter) = begin
```
Comment (ScottPJones) [inline src/BSONIter.jl]: Same here:
`Base.next(obj::BSONObject, state::BSONIter) = ((key(state), value(state)), next!(state))`

File: src/BSONIter.jl [hunk 6]
```diff
@@ -97,35 +98,29 @@ end
...
     ((key(bsonIter), value(bsonIter)), next!(bsonIter))
 end
-export next
 
-done(bsonObject::BSONObject, bsonIter::BSONIter) = begin
+Base.done(bsonObject::BSONObject, bsonIter::BSONIter) = begin
```
Comment (ScottPJones) [inline src/BSONIter.jl]: Same for the rest

File: test/runtests.jl [hunk 7]
```diff
@@ -22,59 +24,64 @@ facts("BSONObject") do
...
+    @test dict(bsonObject) == Dict{Any,Any}("string"=>"Hello, Jérôme","anotherNull"=>nothing,"null"=>nothing,"regularSymbol"=>"symbol","bool"=>true,"int"=>42,"minkey"=>:minkey,"maxkey"=>:maxkey,"binaryData"=>collect(map(t -> UInt8(t), 1:20)), "array"=>Any["hello",(Dict{Any,Any}("foo"=>Any[56,false]))],"subdict"=>(Dict{Any,Any}("key"=>"value")),"double"=>0.125)
     append(bsonObject, "int64", -57)
-    @fact bsonObject["int64"] --> -57
+    @test bsonObject["int64"] == -57
     append(bsonObject, "int32", 0x12345678)
-    @fact bsonObject["int32"] --> 305419896
+    @test bsonObject["int32"] == 305419896
 
-    context("BSONObject Copy Constructor") do
+    # BSONObject Copy Constructor
```
Comment (ScottPJones) [inline test/runtests.jl]: Use nested testsets for context

**Review by ScottPJones — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 2 above)
Comment (felipenoris) [reply .travis.yml]: I usually stick with Codecov.
Do you have both of them set up on this repo?

**Comment by ScottPJones** [comment]
I wouldn't disable the tests for Windows, just use the new appveyor script.
It works great.

File: .travis.yml (hunk 2 above)
Comment (ScottPJones) [reply .travis.yml]: I'll have to check and make sure Codecov is added for both, and get the ids for the badges.
If you look at the README.md for the packages in juliastring.org, you'll see I like to be pretty complete
(but that is for later, not for this PR)

**Comment by felipenoris** [comment]
@ScottPJones Which appveyor script?

**Comment by ScottPJones** [comment]
Check out the one that I use for my juliastring.org packages, such as ChrBase.jl/appveyor.yml.
Note that you can remove all versions except 0.6, since this version of LibBSON.jl is only for v0.6.
Also comment out the "allow_failures" stuff.

File: .travis.yml (hunk 2 above)
Comment (ScottPJones) [reply .travis.yml]: I guess it won't need the part about using Pkg, since this is only for v0.6

**Review by ScottPJones — APPROVED** [review]

**Comment by felipenoris** [comment]
@ScottPJones We're all set with this one! Thanks!

**Pull request merged**